jupyter_server_fileid/manager.py

```diff
@@ -122,9 +132,114 @@ def _copy_recursive(self, from_path: str, to_path: str, path_mgr: Any = os.path)
+        manager_classname = self.__class__.__qualname__
+        self.con.execute(
+            "CREATE TABLE Manifest("
+            "manager_module TEXT NOT NULL, "
+            "manager_classname TEXT NOT NULL"
+            ")"
+        )
+        self.con.execute(
+            "INSERT INTO Manifest (manager_module, manager_classname) VALUES (?, ?)",
+            (manager_module, manager_classname),
+        )
+
+    def _check_manifest(self) -> bool:
+        """
+        Returns whether the database file's manifest matches the expected
+        manifest for this class. If one does not exist, writes the manifest and
+        returns True.
+        """
+        if not hasattr(self, "con") or self.con is None:
+            return False
+
+        manifest = None
+        manager_module = self.__class__.__module__
+        manager_classname = self.__class__.__qualname__
+
+        try:
+            manifest = self.con.execute(
+                "SELECT manager_module, manager_classname FROM Manifest"
+            ).fetchone()
+        except Exception:
+            pass
+
+        if manifest is None:
+            self._write_manifest()
+            return True
+
+        manifest_module, manifest_classname = manifest
+        return manifest_module == manager_module and manifest_classname == manager_classname
+
+    def _migrate_database(self) -> None:
+        """Checks whether the database file's manifest matches the expected
+        manifest for this class. Writes the manifest if one does not exist.
+        If one already exists and is incompatible with the expected manifest,
+        then this method backs up the existing database file into a separate
+        file, and then migrates the database over via calling `export_row()` on
+        the old manager and passing the yielded rows to `self.import_row()`.
+        """
+        if not hasattr(self, "con") or self.con is None:
+            return
+
+        manifest = self.con.execute(
+            "SELECT manager_module, manager_classname FROM Manifest"
+        ).fetchone()
+        prev_module, prev_classname = manifest
+
+        # first, backup the database at backup_db_path.
+        db_path_dir, db_path_basename = os.path.split(self.db_path)
+        backup_db_path = os.path.join(
+            db_path_dir, datetime.datetime.now().strftime("%Y-%m-%dT%H-%M-%S-") + db_path_basename
+        )
+        self.con.close()
+        os.rename(self.db_path, backup_db_path)
+
+        # recreate files table and indices
+        self.con = sqlite3.connect(self.db_path)
+        self._write_manifest()
+        self._create_tables()
+
+        # finally, migrate the database from backup_db_path to the current database.
+        PrevManager = getattr(importlib.import_module(prev_module), prev_classname)
+        for row in PrevManager.export_rows(backup_db_path):
+            self.import_row(row)
+
+    @staticmethod
+    def export_rows(db_path: str) -> Generator[Tuple[str, str], None, None]:
```

Comment:
Migrating a row at a time isn't very performant.  I suggest introducing a configurable _batch size_ and dealing with multiple rows per batch.

Reply:
1. Is there a magic "batch size" that is known to be best for migrating databases? Or is it known that bigger batch sizes are always faster, but you are limited by the memory available on your system?

2. If neither are the case, then users would have to run benchmarks on their own hardware beforehand to get the most value out of this batch size trait. Otherwise, they might configure it to some value that actually makes the migration even slower.

I'm open to setting a default batch size if we can come up with some reasonably justified batch size. However, I can't imagine that users will be migrating their file ID databases frequently enough to warrant a configurable batch size trait. Let's tackle performance concerns later if the need arises and not let them block this PR.

Reply:
Good questions.  There are always trade-offs and "batch-size tuning" is quite common.

> Let's tackle performance concerns later if the need arises and not let them block this PR.

I'm fine with that except that batch size of any value would impact the API and I think we should nail down the API before too long (i.e., the next release).

Could we inject a `List` type hint into the inputs and outputs along with a batch_size configurable that defaults to 1?

```python
    def export_rows(db_path: str) -> Generator[List[Dict]], None, None]:
       ...
```

```python
    def import_rows(self, row: List[Dict]) -> None:
        ...
```
I changed the `Tuple[str,str]` to `Dict` in preparation for returning more columns (next comment).  One side-effect of that, however, is that all of the column names would be redundant on subsequent items in the list.  One approach would be to let the zero-row be a set of tuples of column names and subsequent items be a Tuple of parallel values.

Another approach might be to let the caller indicate the columns (and their order within the "rows" tuples) via a separate tuple, so you'd then have signatures like:
```python
    def export_rows(db_path: str, columns: Tuple) -> Generator[List[Tuple]], None, None]:
       ...
```

```python
    def import_rows(self, columns: Tuple, rows: List[Tuple]) -> None:
        ...
```

Reply:
Yeah, great point regarding the fact this would be a breaking API change. Let's do this as soon as possible, even before we determine a good batch size to default to.

> I changed the Tuple[str,str] to Dict in preparation for returning more columns (next comment). One side-effect of that, however, is that all of the column names would be redundant on subsequent items in the list.

I think the overhead of serializing each row to a dictionary is negligible compared to the performance of disk reads. Furthermore, the `import_rows()` method would have to build a dictionary mapping each column name to the index in the tuple, so we're not really saving on much performance. Let's stick to representing each row as a dictionary rather than as a tuple.

Reply:
Done 👍